### ISSUE TITLE
All Applications which need GLX crash - unable to create glx visual

### ISSUE THREAD

**Issue by ashantia-ai** [apptainer/singularity#5501]
Hi Everyone,

I have created definition file which uses nvidia/cudagl from docker (Cuda version 9.0).
On top of that, I install ROS (robot operating system) and Gazebo (robot simulator), and cudnn. 
Everything runs fine on my own PC, but I have glx issues on the HPC. Please see details below.

### Version of Singularity:

What version of Singularity are you using?
On my PC: 3.5.3
On HPC: version 3.5.3-1.1.el7

### Expected behavior
When I run gazebo, I should see an empty 3d world. This works perfectly on my PC. 
I also tried another ROS application called "rviz" (robot visualization) which again works perfectly on my PC.

### Actual behavior
However, on the HPC, a window pops and dies immediately. The gazebo client error log shows the following:

```
(1597578724 102085341) [Msg] Waiting for master.

(1597578724 106957772) [Msg] Connected to gazebo master @ http://127.0.0.1:11345

(1597578724 107114645) [Msg] Publicized address: 172.23.46.59

(1597578724 108541606) Loaded .ini file from: "/home/p257689/.gazebo/gui.ini"

(1597578724 116882254) [Err] [RenderEngine.cc:714] **Unable to create glx visual**

(1597578724 116913576) [Wrn] [RenderEngine.cc:97] **Unable to create X window. Rendering will be disabled**

(1597578724 116937176) [Wrn] [RenderEngine.cc:301] Cannot initialize render engine since render path type is NONE. Ignore this warning ifrendering has been turned off on purpose.

(1597578724 533394080) [Err] [GLWidget.cc:175] GLWidget could not create a scene. This will likely result in a blank screen.

(1597578724 827831385) [Err] [ModelEditor.cc:75] User camera is NULL. Non-editable models will keep their original material

```

The rviz program crashes with a segmentation fault.

### Steps to reproduce this behavior
This is the definition file

```
Bootstrap: docker
From: nvidia/cudagl:9.0-devel-ubuntu16.04
%files
        ./libcudnn7_7.6.4.38-1+cuda9.0_amd64.deb /opt
        ./libcudnn7-dev_7.6.4.38-1+cuda9.0_amd64.deb /opt
%post
        apt-get -y update && apt-get -y upgrade
        apt-get install -y lsb-release
        # Adding keys for ROS
        sh -c 'echo "deb http://packages.ros.org/ros/ubuntu $(lsb_release -sc) main" > /etc/apt/sources.list.d/ros-latest.list\'
        apt-key adv --keyserver 'hkp://keyserver.ubuntu.com:80' --recv-key C1CF6E31E6BADE8868B172B4F42ED6FBAB17C654
        apt-get update && apt-get install -y ros-kinetic-desktop-full \
                        wget git nano python-pip
        apt install -y python-rosinstall python-rosinstall-generator python-wstool build-essential
        apt install -y vim

        dpkg -i /opt/libcudnn7_7.6.4.38-1+cuda9.0_amd64.deb
        dpkg -i /opt/libcudnn7-dev_7.6.4.38-1+cuda9.0_amd64.deb

        apt-get install -y python-catkin-tools ros-kinetic-move-base
        apt-get install -y ros-kinetic-dwa-local-planner
        /bin/bash -c "curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh | bash"
        apt-get install -y git-lfs software-properties-common
```

You can remove the libcudnn files if you don't have them.

Then:

```
sudo singularity build nav.img nav.def
singularity shell --nv nav.img
Singularity> export DISPLAY=:1.0 && export LANG=C   
Singularity> gazebo
```

### What OS/distro are you running

My PC runs on ubuntu 20.04, and the HPC cluster 

```
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"
CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```

### How did you install Singularity

From the website, I built from source on my PC. I have no idea how was it installed on the HPC cluster. 

### Additional Info
glxgears work on the host. My image does not have glxgears so I cannot test that inside the container.
nvidia-smi works on the host and inside the container.
I have to call a specific command to load modules for the HPC cluster, when I do that, LD_LIBRARY_PATH is updated with a variety of links, I also tried binding the location to a path within the container and update those paths, but it still didn't work.

Does anyone have any clue what I am doing wrong here?

**Comment by jmstover** [comment]
Does the HPC node have a 3D capable graphics card in it (a lot of HPC nodes have a simple 2D card for display)? You're also going to need a Display to render the 3d graphics (running X windows and setup for users to use the display).

Are you running on what would be considered a visualization node?

To check graphics card:
```
lspci | grep -i 'vga'
```

**Comment by ashantia-ai** [comment]
Hey, thanks for the quick reply.
Yeah, it does. I am specifically logging into one of the GPU nodes . The output of your commands give the following:

```
00:0f.0 VGA compatible controller: VMware SVGA II Adapter
02:00.0 VGA compatible controller: NVIDIA Corporation GV100GL [Tesla V100 PCIe 32GB] (rev a1)
```

A display is available, I can run the glxgears program on the host. Or do you mean something else?

**Comment by jmstover** [comment]
> A display is available, I can run the glxgears program on the host. Or do you mean something else?

Nope, that's what I mean. So if you do something like:

```
DISPLAY=:0.0 glxgears
```

From a SSH terminal to the host does it run?

*Edit:* I'm assuming `:0.0` is the root display. And some GPUs can't be used for display... they're pure computation only.

**Comment by ashantia-ai** [comment]
> From a SSH terminal to the host does it run?

Yes. I have two options for host connection. One is SSH which DISPLAY will be :11.0, and another option is through some sort of an interactive X/VNC which the DISPLAY is :1.0, on both cases glxgears works. But 0.0 is not available.

I was wondering, maybe the paths to all the the nvidia libraries are not correct. What does the --nv option actually do? Maybe If I know that, I can manually bind correct paths.

**Comment by jmstover** [comment]
`--nv` takes the host libraries and binds them into the container at `/.singularity.d/libs` ... This path is put at the front of `LD_LIBRARY_PATH`

Do you get the same error with (a SSH connection):

```
SINGULARITYENV_DISPLAY=:11.0 singularity shell --nv nav.img
```

What it _looks_ like is that the container can't make a connection to the X display. As awful as it is, you may want to test with:  `xhost +`  on the host (to allow everything) and see if that works, then check whats allowed to connect until you find what makes it happy. :/

Something like `localuser` should be enough... but it could be blocking socket vs IP connections, etc...

**Comment by ashantia-ai** [comment]
I see, hmm that maybe it. But I tried `xhost +` , that didn't work. The SSH connection gives me another error when running gazebo inside the singularity shell: I think this is a different error and probably related to the fact that I am using SSH. In the SSH connection, I am connecting to a guest node with 2x 1080TIs, while the interactive session is a virtual environment on the main node.

```
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  151 (GLX)
  Minor opcode of failed request:  3 (X_GLXCreateContext)
  Value in failed request:  0x0
  Serial number of failed request:  22
  Current serial number in output stream:  23
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  151 (GLX)
  Minor opcode of failed request:  3 (X_GLXCreateContext)
  Value in failed request:  0x0
  Serial number of failed request:  22
  Current serial number in output stream:  23
```
I don't see the above error using the interactive desktop sessions (X/VNC). Interesting... I am out of ideas

**Comment by jmstover** [comment]
Can you run `glxinfo` ? I'm going to guess not... What this _looks_ like is a clobbering of the nvidia GL libraries vs the mesa GL libraries.

**Comment by ashantia-ai** [comment]
glxinfo on the host? yes, but I don't have it installed on the singularity image. It takes me 5 hours to upload my image to the cluster... slow internet. But I will start uploading and check again.

**Comment by ashantia-ai** [comment]
I think I may have found a fix based on what you said regarding conflict of libraries, after loading the HPC module on the host, when I do `ldd /usr/glxinfo` I get the following:

```
linux-vdso.so.1 =>  (0x00007faceacbd000)
	libGL.so.1 => /software/software/Mesa/19.1.7-GCCcore-8.3.0/lib/libGL.so.1 (0x00007face9f88000)
	libm.so.6 => /lib64/libm.so.6 (0x00007face9c86000)
	libX11.so.6 => /software/software/X11/20190717-GCCcore-8.3.0/lib/libX11.so.6 (0x00007faceab58000)
	libc.so.6 => /lib64/libc.so.6 (0x00007face98b8000)
	libglapi.so.0 => /software/software/Mesa/19.1.7-GCCcore-8.3.0/lib/libglapi.so.0 (0x00007faceab25000)
	libXext.so.6 => /software/software/X11/20190717-GCCcore-8.3.0/lib/libXext.so.6 (0x00007faceab11000)
	libLLVMX86Disassembler.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMX86Disassembler.so.9 (0x00007face9721000)
	libLLVMX86CodeGen.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMX86CodeGen.so.9 (0x00007face935f000)
	libLLVMScalarOpts.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMScalarOpts.so.9 (0x00007face904a000)
	libLLVMInstCombine.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMInstCombine.so.9 (0x00007face8f48000)
	libLLVMTransformUtils.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMTransformUtils.so.9 (0x00007face8d6a000)
	libLLVMX86Desc.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMX86Desc.so.9 (0x00007face8b00000)
	libLLVMX86Info.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMX86Info.so.9 (0x00007faceab0c000)
	libLLVMMCDisassembler.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMMCDisassembler.so.9 (0x00007faceab04000)
	libLLVMMCJIT.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMMCJIT.so.9 (0x00007faceaaf3000)
	libLLVMExecutionEngine.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMExecutionEngine.so.9 (0x00007faceaacc000)
	libLLVMTarget.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMTarget.so.9 (0x00007face8af0000)
	libLLVMRuntimeDyld.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMRuntimeDyld.so.9 (0x00007face8a75000)
	libLLVMBitWriter.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMBitWriter.so.9 (0x00007face8a34000)
	libLLVMAnalysis.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMAnalysis.so.9 (0x00007face8688000)
	libLLVMMC.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMMC.so.9 (0x00007face85ad000)
	libLLVMCore.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMCore.so.9 (0x00007face82bf000)
	libLLVMSupport.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMSupport.so.9 (0x00007face80eb000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007face7ee7000)
	libz.so.1 => /software/software/zlib/1.2.11-GCCcore-8.3.0/lib/libz.so.1 (0x00007face7ece000)
	libstdc++.so.6 => /software/software/GCCcore/8.3.0/lib64/libstdc++.so.6 (0x00007face7d34000)
	libgcc_s.so.1 => /software/software/GCCcore/8.3.0/lib64/libgcc_s.so.1 (0x00007face7d1b000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007face7aff000)
	/lib64/ld-linux-x86-64.so.2 (0x00007faceaa9e000)
	libxcb.so.1 => /software/software/X11/20190717-GCCcore-8.3.0/lib/libxcb.so.1 (0x00007face7ad6000)
	libXau.so.6 => /software/software/X11/20190717-GCCcore-8.3.0/lib/libXau.so.6 (0x00007faceaac2000)
	libXdmcp.so.6 => /software/software/X11/20190717-GCCcore-8.3.0/lib/libXdmcp.so.6 (0x00007face7acf000)
	libLLVMAsmPrinter.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMAsmPrinter.so.9 (0x00007face79f1000)
	libLLVMCodeGen.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMCodeGen.so.9 (0x00007face7514000)
	libLLVMGlobalISel.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMGlobalISel.so.9 (0x00007face7477000)
	libLLVMProfileData.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMProfileData.so.9 (0x00007face7408000)
	libLLVMSelectionDAG.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMSelectionDAG.so.9 (0x00007face712d000)
	libLLVMX86Utils.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMX86Utils.so.9 (0x00007face7125000)
	libLLVMAggressiveInstCombine.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMAggressiveInstCombine.so.9 (0x00007face7112000)
	libLLVMObject.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMObject.so.9 (0x00007face7015000)
	libLLVMBinaryFormat.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMBinaryFormat.so.9 (0x00007face6ff6000)
	libLLVMDebugInfoCodeView.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMDebugInfoCodeView.so.9 (0x00007face6f3d000)
	libLLVMRemarks.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMRemarks.so.9 (0x00007face6f2c000)
	librt.so.1 => /lib64/librt.so.1 (0x00007face6d23000)
	libncurses.so.6 => /software/software/ncurses/6.1-GCCcore-8.3.0/lib/libncurses.so.6 (0x00007face6cc3000)
	libLLVMDemangle.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMDemangle.so.9 (0x00007face6c82000)
	libLLVMDebugInfoDWARF.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMDebugInfoDWARF.so.9 (0x00007face6bf0000)
	libLLVMDebugInfoMSF.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMDebugInfoMSF.so.9 (0x00007face6bdd000)
	libLLVMMCParser.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMMCParser.so.9 (0x00007face6b8f000)
	libLLVMBitReader.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMBitReader.so.9 (0x00007face6b1e000)
	libLLVMBitstreamReader.so.9 => /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/libLLVMBitstreamReader.so.9 (0x00007face6b07000)
```
As you can see the loaded libraries are in this /software directory which gets linked after I load a module. These are **not** taken in automatically with the --nv option. Actually, in this case --nv is taking some wrong libraries. So I went ahead performed a Bind for all these folders and mapped them to /opt/lib1, /opt/lib2 ... /opt/libN, and then gazebo worked.
Although this is strange because the host ENV is pointing to all these by default, but the singularity container ENV was partially different. I will investigate further tomorrow.
```
[X@pg-gpu27 X]$ singularity shell --nv -B ~/.gazebo -B /usr/lib64/nvidia -B /local/tmp -B /software/software/Mesa/19.1.7-GCCcore-8.3.0/lib/:/opt/lib -B /software/software/LLVM/9.0.0-GCCcore-8.3.0/lib/:/opt/lib2 -B /software/software/ncurses/6.1-GCCcore-8.3.0/lib/:/opt/lib3 -B /software/software/GCCcore/8.3.0/lib64/:/opt/lib4 nav.img
Singularity> ldd -v /bin/glxinfo^C
Singularity> export LD_LIBRARY_PATH=/opt/lib:/opt/lib1:/opt/lib2:/opt/lib3:/opt/lib4:$LD_LIBRARY_PATH
Singularity> export LANG=C
Singularity> gazebo
```

Now my question is, is there a less painful way of doing this? ;-)

**Comment by jmstover** [comment]
To  test this ...

```
$ singularity shell --nv nav.img
> echo $LD_LIBRARY_PATH
> export LD_LIBRARY_PATH=/.singularity.d/libs:$LD_LIBRARY_PATH
> export LANG=C
> gazebo
```

The `/.singularity.d/libs` directory should be the first directory

**Comment by ashantia-ai** [comment]
I have tried that before with no success. the output for the `echo $LD_LIBRARY_PATH` in the container is

`/usr/local/lib/x86_64-linux-gnu:/usr/local/lib/i386-linux-gnu:/usr/lib/x86_64-linux-gnu:/usr/lib/i386-linux-gnu:/usr/local/nvidia/lib:/usr/local/nvidia/lib64:/usr/local/nvidia/lib:/usr/local/nvidia/lib64:/.singularity.d/libs`

while the echo in the host is:

`/software/software/Mesa/19.1.7-GCCcore-8.3.0/lib:/software/software/libunwind/1.3.1-GCCcore-8.3.0/lib:/software/software/XZ/5.2.4-GCCcore-8.3.0/lib:/software/software/LLVM/9.0.0-GCCcore-8.3.0/lib:/software/software/libdrm/2.4.99-GCCcore-8.3.0/lib:/software/software/X11/20190717-GCCcore-8.3.0/lib:/software/software/fontconfig/2.13.1-GCCcore-8.3.0/lib:/software/software/util-linux/2.34-GCCcore-8.3.0/lib:/software/software/ncurses/6.1-GCCcore-8.3.0/lib:/software/software/freetype/2.10.1-GCCcore-8.3.0/lib:/software/software/libpng/1.6.37-GCCcore-8.3.0/lib:/software/software/expat/2.2.7-GCCcore-8.3.0/lib:/software/software/bzip2/1.0.8-GCCcore-8.3.0/lib:/software/software/nettle/3.5.1-GCCcore-8.3.0/lib64:/software/software/GMP/6.1.2-GCCcore-8.3.0/lib:/software/software/CUDA/10.1.243-GCC-8.3.0/nvvm/lib64:/software/software/CUDA/10.1.243-GCC-8.3.0/extras/CUPTI/lib64:/software/software/CUDA/10.1.243-GCC-8.3.0/lib64:/software/software/binutils/2.32-GCCcore-8.3.0/lib:/software/software/zlib/1.2.11-GCCcore-8.3.0/lib:/software/software/GCCcore/8.3.0/lib/gcc/x86_64-pc-linux-gnu/8.3.0:/software/software/GCCcore/8.3.0/lib64:/software/software/GCCcore/8.3.0/lib`

Now, If I even manually add the above to the container's path it won't work because for some strange reason the container does not have access to the /software folder. I think the reason was that the permissions are not for the HPC hosts, it is like a mounted directory with a different owner. So I did the following:

```
singularity shell --nv -B /software/software nav.img
> export LD_LIBRARY_PATH=ALL_PATHS_FROM_HOST_COPIED_MANUALLY:$LD_LIBRARY_PATH
> export LANG=C
> gazebo
```

The above works, but even with the new bind, your script won't work. I am quite sure that the --nv option is not taking the correct libs in this case. Is there somewhere a source script that I can see how and what exactly the --nv option takes from the host? I guess there is a library/path conflict there.

**Comment by ashantia-ai** [comment]
Okay, I think I figured it out. Basically the main reason for the problem was indeed the way that the HPC host was dealing with paths. They were using Lmod package to setup and load modules necessary for the environment. So either I had to use the solution above or install Lmod on the container and set it correctly to build the paths. I think I can close the issue now. Thank you @jmstover  for the great support.
